### ISSUE TITLE
iterations of `penprompter` model

### ISSUE THREAD

**Issue by haznai** [haznai/peopen#9]
- design doc: bottleneck the middle input of the future dspy model, that one will be part of the pad, and the source of truth
- [ ] can do data engineering of all the arguments to speed up the process of training!

**Comment by haznai** [comment]
### 2024-09-30

#### Summary
- Achieved initial preprocessing of database into `dspy` examples.
- Initiated `peop-5.py` for the first `dspy` iteration.
- Flowchart created for the first iteration of `dspy` module.
```mermaid
flowchart TD
    Input["Input Data"]
    
    A1["Summarize 'im_detail' (Predictor)\nInputs: 'titel', 'im_detail'"]
    A2["Extract Arguments from Committee (Predictor)\nInputs: 'argumenteKomitee', 'empfehlungKomitee'"]
    A3["Extract Arguments from Federal Council (Predictor)\nInputs: 'argumenteBundesrat', 'empfehlungBundesrat'"]
    
    B1["Clarify Ambiguous Terms (ChainOfThought)"]
    B2["Highlight Legal Requirements (ChainOfThought)"]
    
    Bottleneck["Generate Draft 'Wortlaut' (ChainOfThought)"]
    
    C1["Legal Review (ChainOfThought)"]
    
    Output["Final 'Wortlaut'"]
    
    Input --> A1
    Input --> A2
    Input --> A3
    
    A1 --> B1
    A2 --> B1
    A3 --> B1
    
    B1 --> B2
    
    B2 --> Bottleneck
    
    Bottleneck --> C1
    
    C1 --> Output
```

- Migrated to `dspy` 2.5, no local `dspy` dependency anymore
- Converted all data inputs to `dspy` models to flat strings so they can be processed beter with llms 
- Train and validation data split and saved as pickles.

#### Insights & Challenges
- `til`: Focus on functionality rather than perfecting the code was productive.
- `til`: Streamlined data inputs enhanced usability despite initial complexity.
- Propagation of nested data structures to plain strings simplified modeling but required meticulous parsing.
- Loading data worked seamlessly, but defining data flow within modules remains a bottleneck.

#### Open Questions ui& Next Steps
- [ ] Change the color of suggested ChatGPT/Copilot.
- [ ] Iterate the model on issue #9 (peop-9) and etherpad prototype on issue #5 (peop-5).
- [ ] Focus on a single bottleneck and review design documents for `peop-9` and `peop-5`.
- [x] Host an API for external calls from Etherpad outside Docker.
- [ ] Train the model.
- [ ] Make the necessary calls from Etherpad while ensuring Docker compatibility.

**Comment by haznai** [comment]
### 2024-10-06
#### Summary

- Refactored a model definition, training, and hosting files.
	- Focused on cleaning up the `src` directory by organizing outdated files.
	- new file structure of `src/`:
		- `model_definition.py`: defines all the classes, nothing else
		- `model_training.py`: actul training of the model
		- `model_serving.py`: takes the trained model and hosts it
- Began training runs with `bootstrapfweshotrandomsearch` strategy.
	- gonna stick with `bootstrapfewshowrandomsearch` as it has been proven to work 
		- dspy 2.5 is supposed to work well with typed predictors, but i haven't looked too deep into this
	- Reviewed and improved the module call structure with visualization support.
	- Completed the first successful training run of the new model.
	- `models/2024-10-06_first_run_2024-10-06.json`
- Flowcharts of the were adapted to reflect significant structural changes in the model
```mermaid
flowchart TD
    InputData[Input Data]
    A1[Summarize 'im_detail']
    A2[Extract Arguments from Committee]
    A3[Extract Arguments from Federal Council]
    B1[Clarify Ambiguous Terms]
    B2[Write Human-Understandable Draft]
    C1[Highlight Legal Requirements]
    C2[Generate 'Wortlaut']
    Output[Final 'Wortlaut']

    %% InputData to A1, A2, A3 (arrows not labeled per your instruction)
    InputData --> A1
    InputData --> A2
    InputData --> A3

    %% A1 to B1
    A1 -->|summarized_im_detail| B1

    %% A2 to B1
    A2 -->|arguments_committee| B1

    %% A3 to B1
    A3 -->|arguments_bundesrat| B1

    %% B1 to B2
    B1 -->|clarified_terms| B2

    %% A1 to B2
    A1 -->|summarized_im_detail| B2

    %% A2 to B2
    A2 -->|arguments_committee| B2

    %% A3 to B2
    A3 -->|arguments_bundesrat| B2

    %% B2 to C1
    B2 -->|human_understandable_draft| C1

    %% C1 to C2
    C1 -->|legal_requirements| C2

    %% B2 to C2
    B2 -->|human_understandable_draft| C2

    %% C2 to Output
    C2 -->|final_wortlaut| Output
```

#### Insights & Challenges

- `til`: Avoid using bare `except`; prefer `except Exception` to handle specific exceptions.
- Learned about fixing issues with `typedPredictor` in dspy, version 2.5.2.
- `bertscore` failures were resolved by fixing a race condition (by just running everything in 1 thread

#### Open Questions & Next Steps

- [ ] Optimize `bertscore` evaluation for speed and parallel processing.
- [ ] Rethink and optimize hyperparameters based on the first training run insights.
- [ ] Trace the execution length of module calls and identify bottlenecks with profiling.
- [ ] Review training logs to verify the optimization of rounds/epochs. Ensure more extended training as necessary.
- [ ] Revisit the call structure of modules focusing on metrics integration.
	- [ ] Streamline module chain -> simplify? maybe `clarify ambigous terms` is not really needed?
	- [ ] study by ablation ...
	- [ ] but first, get the training runs right

**Comment by haznai** [comment]
### 2024-10-25

#### Summary
- Incorporated factual consistency language model and retrieval model into `src/model_training.py`.
- The bigger, meta-model is now called `penprompter`
- Rethought and adjusted the model architecture.
<img width="757" alt="Screenshot 2024-10-25 at 11 39 01" src="https://github.com/user-attachments/assets/5e6205db-ba71-4ba0-a8f5-203f8ac60ccf">
- Imported factual consistency model into `penprompter` and began training.
- Achieved a new best score of 67.46 during model training.
	- this is not with the truncated wortlauts!

#### Insights & Challenges
- Encountered issues with module compatibility after changing the architecture:
  - Learned (`til`) that the current setup lacks a clear path to deep copy certain network attributes, leading to unexpected behavior.
- Model scoring shows improvements but needs validation and error-solving on retrieval methods.

#### Open Questions & Next Steps
- [ ] Create an updated network graph with mermaid.
- [ ] Consider writing draft module prompts to argue FOR the initiative.
- [ ] Rethink the evaluation metric to align with the actual `wortlauts`.
- [ ] Check and fix errors in `model_training.py` and address issues with the 'article_number_retriever'.
- [ ] Use [py-readability-metrics](https://pypi.org/project/py-readability-metrics/) for Flesch–Kincaid readability tests.
- [ ] maybe more tests
- [ ] eval final metric with open source llms

**Comment by haznai** [comment]
### 2024-10-26

#### Summary
- Adjusted the `init` signature of the retrieval model to fix the bug with `k` not being present
- Successfully finalized the `penprompter` model without errors after addressing an issue with deep copy warnings; committed the improved model: `models/2024-10-26_penprompter-first-full-training.json`.
- Merged datasets, initiated the creation of a new "blacked out" dataset.
- Shared the model's application on the dspy Discord channel.

#### Insights & Challenges
- **til**: Caching is crucial for efficiency; avoid redundant computations.
- Encountered a warning on deep copying attributes within the `PenPrompterNetwork`; identified the source as the `dspy/primitives/module.py` and confirmed existing awareness and fixes in recent versions, considering potential update if necessary.
	- The lack of `copy` or `deepcopy` functions in `Retrieve(Parameter)` class might be related but isn't currently critical.

#### Open Questions & Next Steps
- [ ] Evaluate the effectiveness of the "blacking out" strategy and merging process of the datasets manually.
- [ ] Train `penprompter` on truncated wortlaut.
- [ ] Evaluate `penprompter` with open-source large language models.

**Comment by haznai** [comment]
### 2024-10-27

#### Summary
- Created a new `final` dataset with truncated wortlauts
	- Ensured correct input keys and created a train/valid split.
- Saved and evaluated new openai model trained with the updated dataset.
- Began experimenting with `together.ai` as a potential OpenAI LLM alternative.
- 
<img width="895" alt="Screenshot 2024-10-28 at 11 19 58" src="https://github.com/user-attachments/assets/b13c49eb-1f1c-4839-9392-0d28ca72d1e0">



#### Insights & Challenges
- `til`: Using `dspy` for prototyping on a 'throwaway' task was insightful; recommendation to do this more often.

#### Open Questions & Next Steps
- [ ] Evaluate `penprompter` with open-source large language models.
- [ ] Generalize the `LanguageModel` overwrite beyond the initial setup.
- [ ] Look at Phoenix logging to see how the output looks like.
- [ ] Try different optimizers for evaluation.

**Comment by haznai** [comment]
### 2025-02-24 

### Summary
- Decided agains modularization in Swift, opting to maintain a single file repository temporarily. 
	- → modularization should be according to the tma
	
- Explored the usage and limitations of `combineLatest`, identifying its restriction to two or three publishers.
- Worked on creating a function to update state asynchronously using `AsyncStream`.

### Insights & Challenges
- `til`: Understanding async algorithms as a wrapper over `asyncsequence`.
- `til`: Discovered that `combineLatest` is best suited for limited publisher streams (two or three).
- `til`: Learned that Tuist separates the declaration and installation phases for dependencies.
- `til`: Gained insights into using continuations in Swift concurrency to manage async states.

### Next
- [ ] develop an asyncstream with all states streaming a la [this article](https://blog.jacobstechtavern.com/p/migrating-combine-to-asyncalgorithms)
- [ ] make an issue for refactoring to modularization according to [tma](https://docs.tuist.dev/en/guides/develop/projects/tma-architecture), if there is not one
